mlem/contrib/kubernetes/base.py

```diff
@@ -0,0 +1,281 @@
+            raise e
+        if host is not None and port is not None:
+            return HTTPClient(host=host, port=port)
+        raise TypeError(
+            f"host and port determined are not valid, received host as {host} and port as {port}"
+        )
+
+
+class K8sEnv(MlemEnv[K8sDeployment]):
+    """MlemEnv implementation for Kubernetes Environments"""
+
+    type: ClassVar = "kubernetes"
+    deploy_type: ClassVar = K8sDeployment
+    """type of deployment being used for the Kubernetes environment"""
+
+    registry: Optional[DockerRegistry] = None
+    """docker registry"""
+    templates_dir: List[str] = []
+    """list of dirs where templates reside"""
+
+    def get_registry(self, meta: K8sDeployment):
+        registry = meta.registry or self.registry
+        if not registry:
+            raise MlemError(
+                "registry to be used by Docker is not set or supplied"
+            )
+        return registry
+
+    def get_image_name(self, meta: K8sDeployment):
+        return meta.image_name or generate_docker_container_name()
+
+    def get_server(self, meta: K8sDeployment):
+        return (
+            meta.server
+            or project_config(
+                meta.loc.project if meta.is_saved else None
+            ).server
+        )
+
+    def deploy(self, meta: K8sDeployment):
+        self.check_type(meta)
+        redeploy = False
+        with meta.lock_state():
+            meta.load_kube_config()
+            state: K8sDeploymentState = meta.get_state()
+            if state.image is None or meta.model_changed():
+                image_name = self.get_image_name(meta)
+                state.image = build_k8s_docker(
+                    meta=meta.get_model(),
+                    image_name=image_name,
+                    registry=self.get_registry(meta),
+                    daemon=meta.daemon,
+                    server=self.get_server(meta),
+                )
+                meta.update_model_hash(state=state)
+                redeploy = True
+
+            if state.deployment_name is None or redeploy:
+                generator = K8sYamlGenerator(
+                    namespace=meta.namespace,
+                    image_name=state.image.name,  # type: ignore
+                    image_uri=state.image.uri,  # type: ignore
+                    image_pull_policy=meta.image_pull_policy,
+                    port=meta.port,
+                    service_type=meta.service_type,
+                    templates_dir=meta.templates_dir or self.templates_dir,
+                )
+                create_k8s_resources(generator)
+
+                if pod_is_running(namespace=meta.namespace):
+                    deployments_list = (
+                        client.AppsV1Api().list_namespaced_deployment(
+                            namespace=meta.namespace
+                        )
+                    )
+
+                    if len(deployments_list.items):
```

Comment:
should we raise error otherwise?

Reply:
`DeploymentError` should be fine i guess? But we also raise it below. We can raise it here also but with a different message.

Reply:
did that

Reply:
Now it's to far apart to be readable :( can you invert conditions?
like
```
if not ...:
   raise ...
... continue with what was under if block
```

Reply:
done